jetty-client/src/main/java/org/eclipse/jetty/client/AbstractHttpConnection.java

```diff
@@ -431,19 +431,20 @@ public boolean cancelIdleTimeout()
 
     protected void exchangeExpired(HttpExchange exchange)
     {
-        synchronized (this)
-        {
-            // We are expiring an exchange, but the exchange is pending
-            // Cannot reuse the connection because the reply may arrive, so close it
-            if (_exchange == exchange)
-            {
-                try
-                {
-                    _destination.returnConnection(this, true);
-                }
-                catch (IOException x)
-                {
-                    LOG.ignore(x);
+        // must lock destination prior to locking *this* to prevent deadlock because:
+        //   1. _destination.returnConnection is synchronized
+        //   2. destination.send is synchronized and calls *this*.send which is synchronized
+        // Destination lock and connection lock must always be locked in the same order
+        synchronized (_destination) {
```

Comment:
One potential problem with this approach is that we will potentially hold the destination lock a lot longer than it otherwise would have been. Taking a look at this excerpt, for example:

https://github.com/databricks/jetty.project/blob/a7a96d65153c48c715a1fec96461c99092f6a968/jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java#L452-L468

The previous code seemed to intentionally keep the `startNewConnection` outside of the synchronized block. Could this cause a kind of unexpected contention if this takes a particularly long time?